cmd/aks-periscope/aks-periscope.go

```diff
@@ -40,8 +42,56 @@ func main() {
 		}
 	}
 
+	collectors, diagnosers, exporters := initializeComponents(creationTimeStamp, hostname)
+
+	//dataProducers includes all selected collectors and diagnosers
 	dataProducers := []interfaces.DataProducer{}
+	for _, c := range collectors {
+		dataProducers = append(dataProducers, c)
+	}
+	for _, d := range diagnosers {
+		dataProducers = append(dataProducers, d)
+	}
+
+	collectorGrp := new(sync.WaitGroup)
+	runCollectors(collectors, exporters, collectorGrp)
+	collectorGrp.Wait()
+
+	diagnoserGrp := new(sync.WaitGroup)
+	runDiagnosers(diagnosers, exporters, diagnoserGrp)
+	diagnoserGrp.Wait()
+
+	zipAndExportString, found := os.LookupEnv("ZIP_AND_EXPORT")
```

Comment:
💡What will be associated `deployment.yaml` changes will look like? For this new var `ZIP_AND_EXPORT`

Reply:
https://github.com/davidkydd/aks-periscope/blob/0071201dd41cd159e87e5a574bdfbcc271c122b1/deployment/aks-periscope-arc.yaml#L164 - minor nitpick but if you want to try find a way that the bool value doesn't have to be wrapped in double quotes that would be preferable

Reply:
I'd suggest not to modify this part as we should remove this zip function asap. What do you think?

Reply:
💡 +1 to what Arnaud is trying to convey, I think `ZIP_AND_EXPORT` is a new design somehow (or Use case probably), like mentioned above when `true` what does it achieve? 

* i.e. in current already re-factored `in-memory` files get uploaded as logs, and `zipped` in the end.

This new flag adds new behaviour. Why is this flag required at all `ZIP_AND_EXPORT`?

Reply:
1. The benefit of removing the zip function is that there is less processing done on the node running periscope.
2. The benefit of **using*** the zip function is that there are less bits needed to be sent over the network.

I agree that the current usage of the zip functionality is not optimized. However as there are benefits in both zipping and not-zipping, instead of removing it we should make it configurable.

Zipping has less benefit when exporting to AzureBlob, because the traffic will usually all be within the MSFT network and hence size is less critical (though still important). 

Arc have asked for a LocalMachineExporter, and the CLI is then used to download periscope outputs directly from the node. This should be zipped first as size is likely more important consideration.

The current function of ZIP_AND_EXPORT is a minimal change which supports the Arc LocalMachineExporter scenario, and allows other clients to still disable the ZIP functionality as an optimization if they wish to just rely on the current behaviour of the AzureBlobExporter. 

Later - we should refactor the AzureBlobExporter so that it also respects the ZIP_AND_EXPORT config (or some equivalent config), which will allow the user to configure the behaviour, and decide which of 1. or 2. is more important for their scenario. In memory files should either be zipped and exported to AzureBlob, or just exported as files - doing both is redundant. The ZIP_AND_EXPORT setting (or whatever we decide to call it) should be configurable for each Exporter, so that users can if they wish, have the AzureBlobExporter exporting un-zipped files, and the LocalMachineExporter exporting a zipped file. Zipping should only be performed once and the file shared by exporters for efficiency.

Reply:
💡 So this refactor is based on some kind of requirements? Thats where we think there is something neither Arnaud and I know or understand, hence the point fo this whole change is not convoluted.

* Currently this tool ==> Runs as daemonset -> collects logs (collector/diagnoser) -> upload it to the azure storage.

In the current state of how this tool works, writing files to in-memory and then zip and export to storage is 1 unit of work. Currently this is what this tool does and code reflects that.

* Why are we moving away from this? 
* Why we are making this configurable? (Are we doing all this to fit any scenario which we don't know about?) 

Unless things are clear why are we adding new configurable flags, like for example only thing configurable before was 1-thing which is `storage account` name and key. Now we have so many moving parts.

Thanks,

Reply:
Both zipping and non-zipping have their advantages and disadvantages, therefore we should support both and expose the configuration necessary for this:

- We know we need a LocalMachineExporter, which should use the ZIP function in order to reduce the amount of data that needs to be sent over the network to the CLI.

- We know we need an AzureBlobExporter, this should probably NOT use the ZIP function because we don't want to increase the amount of processing cycles which periscope consumes on the customer cluster, in the case that it is only outputting to AzureBlob storage and the bits are staying within the microsoft network.

Therefore we know we need to support both ZIP and non-ZIP. 

More generally - declarative config should not be considered as moving parts, it is a flat statement of what we want the tool to do, the moving parts are on the inside baked into the image. Adding additional config isn't something to fear or shy away from - we should look to make periscope configurable and tunable via exposing config values where ever it makes sense, as this provides a great deal of flexibility to adapt or to support new scenarios which we might not have considered. 

If we don't add new configuration flags, then the config is still there - it is just set to a single value and hard-coded into the image.

Reply:
Then take zip code and put it in the localmachineexporter and remove it from there. 
You were speaking about reducing the amount of data sent over the network; today because we have this zip at this location, we are sending a lot more than necessary (because others diagnoser / collectors are already sending the same data)

Reply:
💡Also,  Why do we need `LocalMachineExporter` ? what benchmarking do we have for the amount of processing cycle it uses right now and improvement this will do? re:  `ZIP function because we don't want to increase the amount of processing cycles which periscope consumes on the customer cluster,` 

There are many open ended questions again, like what all goals this PR eventually have.

We handle the in-memory logs and what current mechanism is to make less noise and in the end zip and export, this `LocalMachineExporter` what is the use case for this? Like in what scenario this will happen. Currently this tool don't support it at-all because the pods in each nodes upload these logs.

Reply:
why restrict zip functionality to localMachineExporter? It is a general feature and other exporters in future might wish to make use of it - even the AzureBlobExporter.

Yes we are sending more data than necessary currently - with this config flag other tools / consumers of periscope can choose to disable the zip feature if they wish.

Also note this PR isn't aimed at and doesn't change the current behaviour of the zip (the flag value defaults to true)

Reply:
Arc need LocalMachineExporter for the case where a cluster does not have access to egress to Azure storage due to firewall or other egress restrictions.

We don't need to benchmark - it is a general statement that zip requires CPU and reduces file size

Reply:
💡 Cool, so this PR #57 is somehow connected to this work, i.e. this is ground up for for achieving localMachineExporter? Because if it is for localMachineExporter we should give some thought to the actual use of local node file save vs where_to_in_local_save?

* suggestion: I would recommend to open a WorkItem clearly summarise or detail the purpose of these refactor, probably at this repo level for this and possibly what the end goal for refactor, or this is all-in-one refactor PR?

* Regarding data and other discussion, I would recommend thinking of `client-go` adaptation. I might land something as first adaptation point but any other PR should first think of how we can use `client-go` for various other PR(s) and is it feasible, (or need discussion etc)

Thanks 🙏

Reply:
Not PR #57 exactly but yes - I have an equivalent LocalMachineExporter which is rebased on the latest changes that have since gone into master.

This PR does not change the location of the node file save, if we want to debate or change this we can do this later - outside of this PR.

This is a refactor to enable the current set of requirements which we have discussed many times - I don't see the value in opening new workitems for this work that is already completed?